### DISCUSSION TITLE
Valetudo 2021.06.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#968] [Release Discussions]
<div align="center">
<img src="https://github.com/Hypfer/Valetudo/blob/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2021.06.0</h2></p>
</div>

This release features Swagger UI for proper REST API Documentation.
Also bugfixes and stability + performance improvements like the changelog of every single android app you have installed.

## Swagger UI

![image](https://user-images.githubusercontent.com/974410/123315997-9f21c580-d52c-11eb-8271-3f5ffd81931b.png)

In a tedious and brain-melting process, OpenAPI documentation was added to Valetudo.
By navigating to `ROBOT_IP/swagger/` you now have an interactive overview for the REST API, which directly lets you interface with the robot.
The schemas made for this are also used by the backend to validate every incoming request.

Since staring at JSON Schemas all day is a pretty mind-numbing task, I didn't manage to also add examples for all responses and requests.

I did however add examples for the Timers Endpoint, meaning that it is now easier to use those again.

![image](https://user-images.githubusercontent.com/974410/123316039-ab0d8780-d52c-11eb-9c17-34f20c944c16.png)

UI support of course coming soon.

## Config file Schema Validation

The new schemas are also used to validate the configuration file loaded by Valetudo.
If any errors arise, Valetudo will backup the config and create a new one using the defaults.
This will prevent issues with Valetudo not starting due to invalid configuration data.

The log will tell you what exactly was wrong in your config and where you can find the backup.

### NodeJS v16.4.0

The Runtime was upgraded to v16.4.0 which brings v8 9.1 including the new Sparkplug thingy, which may result in CPU performance improvements.

### MQTT ignores retained messages

A common rookie mistake is that command MQTT messages are sent with the retain flag causing the robot to receive them on every reconnect.
This effectively executed a cleanup on each reboot at 4am.

To combat this, Valetudo will now simply ignore all retained messages received and complain about them in the logfile.

### System REST Endpoints

There are new REST Endpoints providing system statistics such as Memory or CPU usage.

`/api/v2/system/host/info`
```json
{
  "hostname": "rockrobo",
  "arch": "arm",
  "mem": {
    "total": 522792960,
    "free": 358219776,
    "valetudo_current": 59215872,
    "valetudo_max": 59699200
  },
  "uptime": 61036,
  "load": [
    0.255,
    0.2725,
    0.28
  ]
}
```

### Code Compression

Due to the switch to `vercel/pkg` 5.3.0, Valetudo now uses the code compression feature, which results in smaller binaries.

### Memory Usage

MQTT Map compression is now streamed, which may or may not improve memory usage on 256mb ram robots.

### Features

* **mqtt:** Ignore received retained message to work around common user errors ([26d8a4c](https://github.com/Hypfer/Valetudo/commit/26d8a4cd4c36410afabfda6f9e87e08fd01e9a28))
* **mqtt:** Stream map serialization to improve memory usage with large maps ([98b2757](https://github.com/Hypfer/Valetudo/commit/98b2757aec0af94392038577eead7749ef659e17))
* **timers:** Add endpoint which returns timer actions supported by this robot ([fe1fa88](https://github.com/Hypfer/Valetudo/commit/fe1fa88d715e1002d3ee607ce9646b8d8b2cdb9d))
* **timers:** add ValetudoGoToTimerAction ([#953](https://github.com/Hypfer/Valetudo/issues/953)) ([d8a523e](https://github.com/Hypfer/Valetudo/commit/d8a523e40138f679411d2b8c33db3057d5f6c8d8))
* **vendor.roborock:** Add support for the Roborock S7 ([c9ccb4a](https://github.com/Hypfer/Valetudo/commit/c9ccb4a11dc546da82684c5f2a80e96847ff2639))
* **webserver:** Add SystemRouter ([efc46e6](https://github.com/Hypfer/Valetudo/commit/efc46e69001a2533077b1e30fd1d8f6de910e263))
* **webserver:** Normalize reported system load ([b384ac3](https://github.com/Hypfer/Valetudo/commit/b384ac3038857f3e8677bac5859b8c1de99f80be))
* **webserver:** Streamed compression of some heavy requests ([0d7d836](https://github.com/Hypfer/Valetudo/commit/0d7d8364b98e950d186d961062b81fbfff5e920a))
* **webserver:** Validate actual endpoint existence against openApi schema ([f7cee31](https://github.com/Hypfer/Valetudo/commit/f7cee31eb20cba824d3fd2ed47cdb73a242de3f9))
* Swagger Docs ([7842d9e](https://github.com/Hypfer/Valetudo/commit/7842d9ec0957a9a45f6c83ab357a901884b11ba1)), closes [#892](https://github.com/Hypfer/Valetudo/issues/892)
* Validate configuration file and create a new one on error ([39e4613](https://github.com/Hypfer/Valetudo/commit/39e46137872f6774d2ebaf1120a10d0cddfff45b))

### Bug Fixes

* **configuration:** Add migration for invalid mqtt port type ([1220aa5](https://github.com/Hypfer/Valetudo/commit/1220aa5622dbf200ef2be88612f1e503ec887328))
* **miio:** Allow setting new wifi credentials when provisioned ([32a12cf](https://github.com/Hypfer/Valetudo/commit/32a12cf82c698b64b3cc037a869f058fb4588f26)), closes [#657](https://github.com/Hypfer/Valetudo/issues/657)
* **mqtt:** Calling GoTo Presets should use regular strings and not json strings ([ec975aa](https://github.com/Hypfer/Valetudo/commit/ec975aaec40c9857e7b35622253128ef75be3fbe)), closes [#960](https://github.com/Hypfer/Valetudo/issues/960)
* **mqtt:** Use semaphore to avoid reconfigure race condition ([5b4b377](https://github.com/Hypfer/Valetudo/commit/5b4b377f98c5030aa13f2133b5a220c3366fae57))
* **ntpclient:** Reduce loglevel of more error codes to avoid confusion ([ec4ee08](https://github.com/Hypfer/Valetudo/commit/ec4ee082eaba88d51d8948ebc97c0cb6e4347ed1))
* **ui:** Allow resetting map when PersistentMapControl is unavailable ([#954](https://github.com/Hypfer/Valetudo/issues/954)) ([bc8feb9](https://github.com/Hypfer/Valetudo/commit/bc8feb936d373757093b79ed40ccd8b7318e8f67))
* **ui:** Hide defunct zones button ([ae7c059](https://github.com/Hypfer/Valetudo/commit/ae7c05948c9cf7252bfedd1fe935c65d69a986b3))
* **vendor.dreame:** Handle some previously unhandled messages ([810a212](https://github.com/Hypfer/Valetudo/commit/810a21226bb1580ed2afae9e2a2fad2b6faa475d))
* **webserver:** Fix SystemRouter units ([fd908f6](https://github.com/Hypfer/Valetudo/commit/fd908f6da756d3d5a6f8350e1a4f4041990a6cf9))
* **webserver:** os.freemem() isn't what it appears to be ([80743f4](https://github.com/Hypfer/Valetudo/commit/80743f455dd20d2abc316260308b9d322c33e284))
* **webserver:** Properly report memory info for kernels older than 3.14 but newer than 2.6 ([9000f51](https://github.com/Hypfer/Valetudo/commit/9000f51eaf818ca465ad0b2ef9f91ca421ac2da4))

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2021.06.0'>Valetudo 2021.06.0</a>.</em>

**Comment by dmlemos** [comment]
Thanks for another release.

> ui: Hide defunct zones button

Could you clarify what happened to _zones_? why was it removed?

**Reply by Hypfer** [reply]
Calling multiple presets at once has been removed since that didn't make sense conceptually.

See also #952

**Reply by dmlemos** [reply]
Contrary to the bug report, I've never had any issue whatsoever with zones on `2021.04.0`. Instead of removing this feature, wouldn't it be wise to investigate the regression? I say this, because segments don't replace zones. I use zones to send the robot to clean specific places that requires our action, or moving something. E.g. foldable table, garbage location, a trolley, etc. These spots require regular cleaning, and as far as I can see this is simply not possible to do with segments.
Is there an alternative? or I'm forever stuck with an old version? 😟

**Reply by Hypfer** [reply]
Did you read the linked issue in its entirety?

**Reply by dmlemos** [reply]
As far as I can see, the feature is deprecated. The alternative is to draw the zones on the map when required? unfortunately this is quite the work in our case.
Was there something else I missed?

**Reply by gitpel** [reply]
the feature **Zones** was also important for me too (his wife)

**Reply by seuffert** [reply]
ZONES was *the* unique feature why i used valetudo in the first place ... this really makes me sad. Just using the "start" button rebuilds the map every time ...
